neqo-transport/src/recovery.rs

```diff
@@ -209,12 +227,152 @@ impl LossRecoverySpaces {
+        }
+
+        if self.congestion_window < self.ssthresh {
+            // Slow start.
+            self.congestion_window += u64::try_from(pkt.size).unwrap();
+            qinfo!([self], "slow start; cwnd {}", self.congestion_window);
+        } else {
+            // Congestion avoidance.
+            qinfo!(
+                [self],
+                "congestion avoidance; cwnd {} ssthresh {}",
+                self.congestion_window,
+                self.ssthresh
+            );
+            self.congestion_window += u64::try_from(MAX_DATAGRAM_SIZE * pkt.size as usize).unwrap()
+                / self.congestion_window
+        }
+    }
+
+    fn on_packets_lost(
+        &mut self,
+        now: Instant,
+        largest_acked_sent: Option<Instant>,
+        pto: Duration,
+        lost_packets: &[SentPacket],
+    ) {
+        for pkt in lost_packets {
+            assert!(self.bytes_in_flight >= u64::try_from(pkt.size).unwrap());
+            self.bytes_in_flight -= u64::try_from(pkt.size).unwrap();
+        }
+
+        qdebug!([self], "Pkts lost {}", lost_packets.len());
+
+        let last_lost_pkt = lost_packets.last().unwrap();
+        self.on_congestion_event(now, last_lost_pkt.time_sent);
+
+        let in_persistent_congestion = {
+            let congestion_period = pto * PERSISTENT_CONG_THRESH;
+            largest_acked_sent < Some(last_lost_pkt.time_sent - congestion_period)
+        };
+        if in_persistent_congestion {
+            qinfo!([self], "persistent congestion");
+            self.congestion_window = MIN_CONG_WINDOW;
+        }
+    }
+
+    fn on_packet_sent(&mut self, size: usize) {
+        self.bytes_in_flight += u64::try_from(size).unwrap();
+        qdebug!(
+            [self],
+            "Pkt Sent len {}, bif {}, cwnd {}",
+            size,
+            self.bytes_in_flight,
+            self.congestion_window
+        );
+        assert!(self.bytes_in_flight <= self.congestion_window);
+    }
+
+    fn in_congestion_recovery(&self, sent_time: Instant) -> bool {
+        self.congestion_recovery_start_time
+            .map(|start| sent_time <= start)
+            .unwrap_or(false)
+    }
+
+    fn on_congestion_event(&mut self, now: Instant, sent_time: Instant) {
+        // Start a new congestion event if packet was sent after the
+        // start of the previous congestion recovery period.
+        if !self.in_congestion_recovery(sent_time) {
+            self.congestion_recovery_start_time = Some(now);
+            self.congestion_window /= 2; // kLossReductionFactor = 0.5
+            self.congestion_window = max(self.congestion_window, MIN_CONG_WINDOW);
+            self.ssthresh = self.congestion_window;
+            qinfo!(
+                [self],
+                "Cong event -> recovery; cwnd {}, ssthresh {}",
+                self.congestion_window,
+                self.ssthresh
+            );
+        } else {
+            qdebug!([self], "Cong event but already in recovery");
```

Comment:
Is this worth a debug_assert!() instead?

Reply:
I don't think so, this is called every time there is a lost packet, which could certainly happen while in congestion recovery, couldn't it?